forte/data/data_store.py

```diff
@@ -835,18 +859,97 @@ def _create_new_entry(
+
+    def _validate_entry(self, new_entry: List) -> List:
+        type_name = new_entry[constants.ENTRY_TYPE_INDEX]
+
+        # Handling creation of annotation type entries
+        if self._is_annotation(type_name):
+            begin_idx = self.get_datastore_attr_idx(type_name, "begin")
+            end_idx = self.get_datastore_attr_idx(type_name, "end")
+
+            # If the begin and end index is not set
+            # Data Store will not be able to add this
+            # entry to the Sorted List (the format in which
+            # annotation are stored)
+            if new_entry[begin_idx] is None or new_entry[end_idx] is None:
+                raise KeyError(
+                    "In order to create Annotation or"
+                    "AudioAnnotation, begin and end parameter must be"
+                    "passed in attribute_data."
+                )
+
+            # If payload is not set, set it to its default value,
+            # ie. 0.
+            payload_id_idx = self.get_datastore_attr_idx(
+                type_name, "payload_idx"
+            )
+
+            if new_entry[payload_id_idx] is None:
+                new_entry[payload_id_idx] = 0
+
+        elif any(
+            issubclass(get_class(type_name), cls)
+            for cls in (Group, MultiPackGroup)
+        ):
+            members_idx = self.get_datastore_attr_idx(type_name, "members")
+            type_idx = self.get_datastore_attr_idx(type_name, "member_type")
+
+            if new_entry[members_idx] is None:
+                new_entry[members_idx] = []
+
+            if not issubclass(new_entry[type_idx], Entry):
+                raise ValueError(
+                    "Attributes required to create Group "
+                    "entry are not set correctly."
+                )
+
+            new_entry[type_idx] = get_full_module_name(new_entry[type_idx])
+
+        elif any(
+            issubclass(get_class(type_name), cls)
+            for cls in (Link, MultiPackLink)
+        ):
+            parent_idx = self.get_datastore_attr_idx(type_name, "parent_type")
+            child_idx = self.get_datastore_attr_idx(type_name, "child_type")
+
+            if new_entry[parent_idx] and new_entry[child_idx]:
+                if not issubclass(
+                    new_entry[parent_idx], Entry
+                ) or not issubclass(new_entry[child_idx], Entry):
+                    raise ValueError(
+                        "Attributes required to create Link"
+                        "entry are not set correctly."
+                    )
+
+                new_entry[parent_idx] = get_full_module_name(
+                    new_entry[parent_idx]
+                )
+                new_entry[child_idx] = get_full_module_name(
+                    new_entry[child_idx]
+                )
+
+        return new_entry
+
     def add_entry_raw(
         self,
         type_name: str,
-        attribute_data: List,
-        base_class: Type[Entry],
+        attribute_data: Dict,
```

Comment:
Actually the use of a `Dict` here defeats the purpose of `DataStore`. DataStore is designed to efficiently handle data structures, and adding data should be list minimum array operations. But using Dict adds some additional speed lost. Further, this actually changes the interface types.

Reply:
Okay, I will take this into account